UTF8 Documentation rework

kellertuer opened the pull request:
Works on #101 and introduces a notation overview. Here's the current details

We can not to anything against 

* `\mathrm{M}`, `\mathrm{H}`, `\mathrm{T}` (for the Minkowski-norm, Hermitian and transposed vectors),...
* the same for `\operatorname`s
* \lvert and \rvert (as well as \lVert and \rVert) are hard to replace, since the utf8 symbols (| and ‖) have in TeX a different spacing


I replaced
* all manifolds (M, \mathcal M) in math mode with ℳ
* \langle, \rangle with their utf8s ⟨⟩
* \cdot with
* \ldots with 
* \times with ⨉ 

I noticed that the links to `allocate, `allocate_result` and `number_eltype`return warnings – are they properly exported in `ManifoldsBase`?

Within the table on notations.md on my machine it seems random whether math is recognized or not... can someone verify that? My change from `$` to ``` `` ``` dioes only randomly take effect.

What do you think to the proposed notation? Some of that is really not unified currently (especially dot/inner and (co)tangents).

kellertuer commented:
A first note is – despite the `notation.md` rendering all renders fine on safari. Concerning the unified notation, I'd like some feedback on what we agree – document that in `notation.md` and then unify the decstrings accordingly.

mateuszbaran commented on src/statistics.jl:
`````diff
@@ -560,8 +560,8 @@ var(M::Manifold, x::AbstractVector; kwargs...) = mean_and_var(M, x; kwargs...)[2
 
-````math
-\sqrt{\frac{1}{c} \sum_{i=1}^n w_i d_{\mathcal M}^2 (x_i,m)},
+````math𝒩
`````
This is probably not correct.

kellertuer replied:
Thanks, I missed that one. Will correct that.

mateuszbaran commented:
It looks quite nice in github file preview. I'll check it more closely later.

> I noticed that the links to `allocate, `allocate_result`and`number_eltype`return warnings – are they properly exported in `ManifoldsBase`?

`allocate` and `number_eltype` are exported in both `ManifoldsBase` and `Manifolds`. I don't know why there are warnings.

> What do you think to the proposed notation? Some of that is really not unified currently (especially dot/inner and (co)tangents).

It's OK. One small issue is that I'm used to denoting tangent vectors by lower case letters (u, v, w) but I can use your notation here.

kellertuer commented:
> It's OK. One small issue is that I'm used to denoting tangent vectors by lower case letters (u, v, w) but I can use your notation here.

i am not saying my proposal should definetly stay unchanged, it's just close to what I am used to (despite that we currently write `( , )_x` for the inner product since theres also a dual pairing between tangent and cotangent). I do think it's nicer to have different sets for points & tangents.

sethaxen commented on docs/src/notation.md:
```diff
@@ -0,0 +1,24 @@
+when referring to that code. This is for example the case the caligraphic symbols.
+
+Within the documented functions the utf8 symbols are used whenever possible,
+as long as that renders still in TeX within this documentation.
```
I think this works
```suggestion
as long as that renders still in $\TeX$ within this documentation.
```

sethaxen commented on docs/src/notation.md:
```diff
@@ -0,0 +1,24 @@
+
+| Symbol | Description | Also used | Comment |
+|:--:|:-------------- |:--:|:--- |
+| ``T^*_x ℳ`` | The cotangent space at ``x`` | | |
```
Either way, doesn't render correctly for some reason: see https://julianlsolvers.github.io/Manifolds.jl/previews/PR104/notation.html

kellertuer replied:
That table took me quite some time, neither `$`, _nor_ ` `` ` seem to render correctly within tables :/

sethaxen replied:
Hmm that's weird. I don't have time to troubleshoot this right now, but I once got math working successfully in tables, but maybe it was a fluke. See https://raw.githubusercontent.com/arviz-devs/ArviZ.jl/master/docs/src/api.md. If that doesn't help, then it might help to check slack or open a Documenter.jl issue.

kellertuer replied:
I have not yet checked (exercise in 10 minutes, yes that's around 7:20 am 😴 ) but it might be that when there is no space between `|` and ` `` ` some rendering gets confused or something. Will investigate during the day (hopefully).

sethaxen commented on docs/src/notation.md:
```diff
@@ -0,0 +1,24 @@
+| ``T^*_x ℳ`` | The cotangent space at ``x`` | | |
+| ``ξ`` | A cotangent vector from ``T^*_x ℳ`` | ``ξ_1,ξ_2,…,η,𝛇`` | |
+| ``F`` | A fiber | |
+| ``⟨·,·⟩`` | inner product (in ``T^*_x ℳ``) | ``⟨·,·⟩_x, g_x(·,·)`` |
```
Can you verify that `⟨` looks the same as `\langle` when rendered, including spacing? Looks a bit cramped in the docs.

sethaxen replied:
You mean `T_x ℳ` right? The tangent space?.
The notation on the left and right are different.
There's also Einstein notation which is convenient for some of the metric functions, but I don't think we need to document that here.

kellertuer replied:
Well it looked cramped because it was not TeXed, see comment below for markdown files and within tables let's got back to TeX instead of utf8. Oh and yes the inner product is on ´T_xM` of course.

sethaxen commented on docs/src/notation.md:
```diff
@@ -0,0 +1,24 @@
+| ``𝒫_{y\gets x}X`` | parallel Transport |
+| ``x`` | A point on ℳ | ``x_1,x_2,…,y,z`` | |
+| ``T_x ℳ`` | The tangent space at ``x`` | | |
+| ``X`` | A tangent vector from ``T_x ℳ`` | ``X_1,X_2,…,Y,Z`` | sometimes with the base point in the index, ``X_x``. |
```
This upper case seems to be a new notation convention. Are we sure we want to change it everywhere?

kellertuer replied:
That is my proposal of a notation. We should agree (from that list) for a notation an I'll unify it. Currently, we use about 3 or maybe even 4 different naming conventions/notations.

sethaxen commented on src/groups/group.jl:
```diff
@@ -237,7 +237,7 @@ end
 
 Inverse $x^{-1} ∈ G$ of an element $x ∈ G$, such that
-$x \circ x^{-1} = x^{-1} \circ x = e ∈ G$.
+$x ◦ x^{-1} = x^{-1} ◦ x = e ∈ G$.
```
I tried unicode here earlier, but it ended up looking really cramped. See https://julianlsolvers.github.io/Manifolds.jl/previews/PR104/manifolds/group.html#Base.inv-Tuple{Manifold,Any}

kellertuer replied:
I have to check, I think there is a middle-circle-symbol that is not as small as the celsius circle thingy.

sethaxen commented on src/groups/special_euclidean.jl:
````diff
@@ -21,7 +21,7 @@ SemidirectProductGroup(Tn, SOn, RotationAction(Tn, SOn))
 ```
 
 Points on $\mathrm{SE}(n)$ may be represented as points on the underlying product manifold
-$\mathrm{T}(n) \times \mathrm{SO}(n)$ or as affine matrices with size `(n + 1, n + 1)`.
````
This use of `⨉` is also a little cramped.

kellertuer replied:
What should we do here then? It already has both spaces so I hoped it would render fine in both?

sethaxen commented on src/manifolds/CholeskySpace.jl:
`````diff
@@ -191,11 +191,11 @@ Parallely transport the tangent vector `v` at `x` along the geodesic to `y`
 ````
 
-where $\lfloor\cdot\rfloor$ denotes the strictly lower triangular matrix,
+where $\lfloor·\rfloor$ denotes the strictly lower triangular matrix,
`````
Maybe `⌊` and `⌋`

kellertuer replied:
Will use them, too, yes.

sethaxen commented on src/manifolds/Hyperbolic.jl:
```diff
@@ -1,31 +1,31 @@
     Hyperbolic(n)
 
-Generate the $\mathbb H^{n}\subset \mathbb R^{n+1}$
+Generate the $ℍ^{n}\subset ℝ^{n+1}$
```
```suggestion
Generate the $ℍ^{n} ⊂ ℝ^{n+1}$
```

kellertuer commented:
For the mathcals, `\times` and `\circ` I will change that back then. Yes, I was also not so happy with those after changing them.

kellertuer commented on docs/src/manifolds/product.md:
```diff
@@ -1,6 +1,6 @@
 # Product manifold
 
-Product manifold $M = M_1 \times M_2 \times \dots M_n$ of manifolds $M_1, M_2, \dots, M_n$.
+Product manifold $M = M_1 ⨉ M_2 ⨉ \dots M_n$ of manifolds $M_1, M_2, \dots, M_n$.
```
I already did that for `\ldots`, which is what I think you unicoded in your comment, but there's also the ones centered t can use for `\dots`. Will change that.

sethaxen replied:
At least for Atom, `\ldots` and `\dots` evaluate to the same unicode.

kellertuer replied:
Yes, I checked that and unified all to that unicode.

kellertuer commented on docs/src/manifolds/product.md:
```diff
@@ -1,6 +1,6 @@
 
-Product manifold $M = M_1 \times M_2 \times \dots M_n$ of manifolds $M_1, M_2, \dots, M_n$.
+Product manifold $M = M_1 ⨉ M_2 ⨉ \dots M_n$ of manifolds $M_1, M_2, \dots, M_n$.
 Points on the product manifold can be constructed using [`ProductRepr`](@ref) with canonical projections $\Pi_i \colon M \to M_i$ for $i \in 1, 2, \dots, n$ provided by [`submanifold_component`](@ref).
```
The trick is, in LaTeX, `\colon` has a different spacing than `:`, with MathJax they are handled equally as far as I read and saw.

kellertuer replied:
Just to clarify, you wrote in the general comment to not use `x` and here you propose to do it?

sethaxen replied:
Here it was a comment RE consistency. It should either be used globally, not used globally, or used globally wherever indistinguishable from `\times`.

kellertuer replied:
The disadvantage of `x` in math environments from TeX is, that it gets set in italics. So I would propose to never use it. if `\times` (I mean its utf8, just too lazy to copy from somewhere ;)) is too cramped in REPL, we can still add spaces fo `  \times `, which would be taken away in TeX anyways and in REPL it looks less cramped.

kellertuer commented on docs/src/manifolds/symmetricpositivedefinite.md:
````diff
@@ -3,7 +3,7 @@
 
 ```math
-\mathcal P(n) = \bigl\{ A \in \mathbb R^{n\times n}\ \big|\ A = A^{\mathrm{T}} \text{ and } x^{\mathrm{T}}Ax > 0 \text{ for } 0\neq x \in\mathbb R^n \bigr\}
+\mathcal P(n) = \bigl\{ A \in ℝ^{n⨉ n}\ \big|\ A = A^{\mathrm{T}} \text{ and } x^{\mathrm{T}}Ax > 0 \text{ for } 0\neq x \inℝ^n \bigr\}
````
Hm, in the general comment you wrote the mathcals are too slanted and yet here you propose to use them? Shall we or shall we not use them?

sethaxen replied:
Similar comment to `\times`. They should be consistently used or not used. This comment was RE consistency, but I prefer they not be used.

kellertuer replied:
I would prefer to consistently use them, just for `\circ` I am not happy with the size of circ in REPL yet, spacing in REPL can always be done with double spaces if needed, in TeX that renders the same.

kellertuer commented:
Sometimes markdown (especially tables) math and utf8 within that seems to work not that nicely. Then I remembered that for `notations.md` we never show that in REPL anyways, so for the markdown files I propose to use classical LaTeX and just for the docstrings the UTF8

kellertuer commented:
Further, I would like to unify notations, the table is my proposal of what I noticed, feel free to suggest further notations and check whether you agree with my proposal. We can of course discuss, With a finished unified notation I would adapt the notation accordingly in the docstrings, where we currently have 3-4 different notations.

kellertuer commented:
I started reworking the docstrings, starting alphabetically with `CholeskySpace` for the manifolds.

mateuszbaran commented on docs/src/notation.md:
```diff
@@ -0,0 +1,22 @@
+as long as that renders still in $\TeX$ within this documentation.
+
+| Symbol | Description | Also used | Comment |
+|:--:|:-------------- |:--:|:--- |
```
I think it could be useful if this table contained Unicode symbols where they are supposed to be used now (like `\mathcal M`) so they can be copied and pasted when needed.

kellertuer replied:
Yes, I tried that, in Markdown that changed quite a few math formulas not to render (with ` `` ` delimiter they were transformed to `$` delimiters, with `$` delimiters they stayed unchanged. I think some UTF8 characters break MathJax or something. So to keep this table unified, I transformed all to just math until I see what the reason is. Also $\langle$ was rendered here rather strangely when using UTF8 (while in docstrings it seems to be nice and right).

mateuszbaran commented on docs/src/manifolds/vector_bundle.md:
```diff
@@ -1,7 +1,7 @@
-Vector bundle $E$ is a manifold that is built on top of another manifold $M$ (base space).
-It is characterized by a continuous function $\Pi \colon E \to M$, such that for each point $x \in M$ the preimage of $x$ by $\Pi$, $\Pi^{-1}(\{x\})$, has a structure of a vector space.
+Vector bundle $E$ is a manifold that is built on top of another manifold $ℳ$ (base space).
+It is characterized by a continuous function $Π : E → ℳ$, such that for each point $x  ∈ ℳ$ the preimage of $x$ by $Π$, $Π^{-1}(\{x\})$, has a structure of a vector space.
```
```suggestion
It is characterized by a continuous function $Π : E → ℳ$, such that for each point $p  ∈ ℳ$ the preimage of $p$ by $Π$, $Π^{-1}(\{p\})$, has a structure of a vector space.
```

kellertuer replied:
I haven't gotten to the vector-bundle yet with renaming, I am in `E` as Euclidean – but I am also carefully checking writing and grammar.

mateuszbaran replied:
OK, sorry for being too hasty about that.

kellertuer replied:
Nevermind. Unifying is a little manual work currently – so it's also good to know what to also look for. I had the `\xi` for SPDs already in mind (I might just `a` instead), but not yet this one.

mateuszbaran commented on src/manifolds/SymmetricPositiveDefinite.jl:
```diff
@@ -1,27 +1,27 @@
-x \in \mathbb R^{n\times n} :
-\xi^\mathrm{T}x\xi > 0 \text{ for all } \xi \in \mathbb R^{n}\backslash\{0\}
+x  ∈ ℝ^{n\times n} :
+ξ^\mathrm{T}xξ > 0 \text{ for all } ξ ∈ ℝ^{n}\backslash\{0\}
```
ξ is now a symbol for cotangent vectors so using it here may be confusing.

mateuszbaran replied:
Similarly it's used in many places in documentation of vector bundles denoting any vector.

kellertuer commented:
@mateuszbaran Is this a bug in https://github.com/JuliaNLSolvers/Manifolds.jl/blob/master/src/manifolds/VectorBundle.jl#L160 ? There is no field `fiber`. After redoing the docs I would like to refactor the bundle to have speaking field names. What do you think about `M.manifold` and `M.vectorspace`? The first would be consistent with other fields.

I will also add a test for above line then.

mateuszbaran commented:
I'm not even sure if that method of `base_manifold` for `VectorSpaceAtPoint`  makes sense actually. I'm not sure in what circumstances it is advisable to call it on an object of type `VectorSpaceAtPoint`.


> Ah, I noticed my mistake – still with this line correct, I would like to rename `M` to `manifold` as a field.

Yes, that's a good idea. I think the field name `fiber` is OK for `VectorSpaceAtPoint`.

kellertuer commented:
Yes, `fibre` is ok, bur we have the field `M` that will get `manifold` and I think the basis has a field `v` – that should be `vectors`, I think.

mateuszbaran commented:
Yes, renaming `M` to `manifold` is a good idea. The other thing you mean is probably the `VS` field of `VectorBundle`? That could also be renamed, to `fiber`, `fibers` or something.

kellertuer commented:
Yes, the `VS` field; `fibre sounds great.

kellertuer commented:
While refactoring the groups I noticed two things

* in general its quite repetitive, especially in `groups.jl`
* I thought we aimed for having the non-mutating (`compose`, `apply_inverse`) etc to just fall back to `similar` and then calling the mutating one. This seems to not happen that much in groups. Is there a reason for implementing both so often?

kellertuer commented:
So this is nearly done. I haven't rewritten all tests, but the rest is done.

Arrays of points are still called `x`, arrays of weights `w`, arrays of vectors `v`. We could rename those to `points`, `weights` and `vectors` for readability

* I would like to check for above stated question on `groups`, whether the THTT can be done “a level higher” and whether all non-mutating ones are necessary
* I would like to carefully read the docs once more.

mateuszbaran commented:
> * in general its quite repetitive, especially in `groups.jl`

That's quite unfortunate but I don't think it can be significantly simplified without a serious effort.

> * I thought we aimed for having the non-mutating (`compose`, `apply_inverse`) etc to just fall back to `similar` and then calling the mutating one. This seems to not happen that much in groups. Is there a reason for implementing both so often?

Yes, there is. First, we may want to eventually provide a truly non-mutating interface for better compatibility with reverse mode AD (especially Zygote, see #42 ). As a result, when defining a wrapper, both variants should be independently forwarded/implemented. Also I believe `Identity` may cause some trouble with the allocate-and-mutate pattern.

kellertuer commented:
I worked through all of those and it felt like at least the THTT could already be done on the `AbstractGroup`-level, if I see how to do that faster, I'll add it here. I think the same can be done for the non-mutating version? Dispatch would still pass that correctly to the right inner mutating one (I think). Will try that the next days.

mateuszbaran commented:
I'm not quite sure but it should be possible to simplify that code. What do you mean by doing THTT on the `AbstractGroup` level? There are many tricky things to consider like a group manifold wrapped in something that is not an `AbstractGroupManifold` but should still act like one. There may also be some methods that need to be defined due to ambiguity errors.

kellertuer commented:
Well currently we do the following for about 10 (2x5) functions for just – I don't know let's take `inverse_translate`. For `exp` we do one function for the non-mutating and the rest on the mutating, which would reduce that to 6 (1+5). Which we could do for nearly all functions defined in `group.jl` (well fo `identity`we actually do that already).

The one thing I am not so sure is, whether we can do the THTT just one and not 16 times within that file. Maybe `base_manifold` does not do the trick because it unwraps all decorators? We could at least to an `undecorate_once` (name to be found still), in order to reduce these 16 THTT to a global one.

kellertuer commented:
Hm, so the idea I tried does not work. I tried

````
base_manifold(M::Manifold, depth::Float64=Inf) = base_manifold(M, is_decorator_manifold(M), depth)
function base_manifold(M::Manifold, ::Val{true}, depth::Float64=Inf)
    return (depth > 0) ? base_manifold(M.manifold, depth-1) : M
end
base_manifold(M::Manifold, ::Val{false}, depth::Float64=Inf) = M
````
to have an “underrate only one” – maybe that's interesting in its own but I fear one still needs 20+ lines just to define `inverse_translate` and its mutating one, because this does not work.

Last idea: Can we adopt the idea from `exp` and just define `exp!` the long way, namely the non-mutating one allocates a variable and calls the mutating one?

mateuszbaran commented:
I think @sethaxen is the most knowledgeable in this area.

By the way, making depth a floating point number is quite non-intuitive. Did you do that to have an actual infinity represented instead of, for example, -1 or a very big constant?

kellertuer commented:
Yes, maybe he can help on whether we sould switch to the scheme we do for the classical functions.

Or there was no specific reason besides using `Inf` and just shortly trying that out. I even just commented out the import of the original ffrom Base to test that. We could also use a large int and docuemt that well.

sethaxen commented:
> > * I thought we aimed for having the non-mutating (`compose`, `apply_inverse`) etc to just fall back to `similar` and then calling the mutating one. This seems to not happen that much in groups. Is there a reason for implementing both so often?
> 
> Yes, there is. First, we may want to eventually provide a truly non-mutating interface for better compatibility with reverse mode AD (especially Zygote, see #42 ). As a result, when defining a wrapper, both variants should be independently forwarded/implemented. Also I believe `Identity` may cause some trouble with the allocate-and-mutate pattern.

Yes, ultimately we need an AD-compatible approach. The safest such approach is to implement both mutating and non-mutating, but there's a chance we can get around that with some way of annotating when a variable is being only mutated or when it's being used as input for operations (for Zygote, at least, these cannot happen at the same time).

The group functions are a bit unique in that in general they are tied for the moment to a single abstract type, `AbstractGroupManifold`. However, one can decorate a group, and it should still be a group. For the usual `AdditionOperation` and `MultiplicationOperation` cases, we can easily implement the non-mutating versions e.g. of `compose` using the non-mutating `*`. So the forwarding pattern for e.g. `inverse_translate` will always forward to `inverse_translate` on the decorated manifold until an implementation is encountered. If a group doesn't define a custom non-mutating version, then this behaves the same as the classical version. If it does though, then this avoids unnecessary mutation.

> Yes, maybe he can help on whether we sould switch to the scheme we do for the classical functions.

What are the problems with the existing implementation? And especially with respect to the docs rework?

kellertuer commented:
Oh indeed there is nothing wrong with the docs, so this might be the wrong place. Let's discuss that somewhere else. I just felt in unifying notations that that part is a lot of stuff the same and the same again (and about 8-10 function definitions per function). 

But you'Re right, I have no expierience what we need, that's why I carefully asked – is all that verbose defining necessary? I tried a short approach, but I don't see directly how to shorten that, despite going the same way as with `exp` and `exp!` (i.e. `exp` internally calls `exp!` and for the functions on groups we just have to make THTT once not twice per function). But If I understood you correctly, that's not possible here?

Anyways, let's discuss that somewhere where it fits – and for now just keep in mind that I feel these 800 lines to be quite repetitive.

Overall I think this Is ready to merge.

sethaxen commented:
> Oh indeed there is nothing wrong with the docs, so this might be the wrong place. Let's discuss that somewhere else. I just felt in unifying notations that that part is a lot of stuff the same and the same again (and about 8-10 function definitions per function).
> 
> But you'Re right, I have no expierience what we need, that's why I carefully asked – is all that verbose defining necessary? I tried a short approach, but I don't see directly how to shorten that, despite going the same way as with `exp` and `exp!` (i.e. `exp` internally calls `exp!` and for the functions on groups we just have to make THTT once not twice per function). But If I understood you correctly, that's not possible here?

Well it's a bit comparing apples to oranges. There are so many functions because of the forwarding. Forwarding the non-mutating just doubles the number of functions as you say. `exp` and `exp!` can be so simple because they don't forward by default. So the real question maybe is whether there's a better way to do forwarding (i.e. with a macro or some utility function). I don't follow the THTT reference (I don't think what we're currently doing is technically THTT).

#84 might be the right issue to continue this discussion.

> Overall I think this Is ready to merge.

Okay, I'll try to review once more today.

kellertuer commented:
Just as a side remark: We are not going into a private function with the extra parameter to dispatch on, but despite that our function dispatch patterns are quite closely following THTT ( https://github.com/JuliaLang/julia/issues/2345#issuecomment-54537633 ).

sethaxen commented:
I just looked through the [preview docs](https://julianlsolvers.github.io/Manifolds.jl/previews/PR104/index.html) and made some notes. First, the notation table is really inconsistent with capitalization.

A few more conventions that we can discuss and maybe should go in the table:
- `𝔤` for Lie algebra
- We should probably also adopt `\mathcal{G}` for a group.
- Then there's `G`, the matrix representation of a metric at a point.
- I propose `𝔽` consistently for a number system type (currently we use `F` or `T` sometimes). If it's a variable, the variable should be `field` (even though it's not always a field, it usually will be, and that makes it clear what it is).
- geodesics are `γ`
- Perhaps transpose and hermitian transpose should both be in the table, since we use `\mathrm` for those.
- `d` as Riemannian distance
- For dimension, we also sometimes use the notation `\dim \mathcal{M} ≡ \dim_{ℝ} \mathcal{M}` to indicate the dimension (equivalently real dimension) of the manifold or number system.
- Our differential notation from `Rotations` and the group manifolds
- Our translation notation `τ` from group manifolds
- `\operatorname{Exp}`/`\operatorname{Log}` as matrix exp/log

I also noticed a few docstrings that are currently mangled (not necessarily from this PR):
- https://julianlsolvers.github.io/Manifolds.jl/previews/PR104/manifolds/sphere.html#ManifoldsBase.check_tangent_vector-Union{Tuple{N},%20Tuple{Sphere{N},Any,Any}}%20where%20N
- https://julianlsolvers.github.io/Manifolds.jl/previews/PR104/manifolds/power.html#ManifoldsBase.manifold_dimension-Union{Tuple{PowerManifold{#s79,TSize,TPR}%20where%20TPR%3C:AbstractPowerRepresentation%20where%20#s79%3C:Manifold},%20Tuple{TSize}}%20where%20TSize

There are also still a lot of unicode script characters for manifold still in the docs that should be switched back to `\mathcal`.

I'm going to do a review of the code now.

sethaxen commented on src/manifolds/Circle.jl:
```diff
@@ -16,178 +16,176 @@ struct Circle{F} <: Manifold where {F<:AbstractNumbers} end
 
-For the complex-valued case the formula is the same as for the [`Sphere`](@ref)
-applied to valuedin the complex plane.
+For the complex-valued case the formula is the same as for the [`Sphere`](@ref) $𝕊^1$ is applied, to values in the
```
```suggestion
For the complex-valued case, the same formula as for the [`Sphere`](@ref) $𝕊^1$ is applied to values in the
```

sethaxen commented on src/manifolds/Circle.jl:
```diff
@@ -196,179 +194,176 @@ function inverse_retract(M::Circle, x::Number, y::Number, ::LogarithmicInverseRe
 
-For the complex-valued case the formula is the same as for the [`Sphere`](@ref)
-applied to valuedin the complex plane.
+For the complex-valued case the formula is the same as for the [`Sphere`](@ref) $𝕊^1$ is applied, to values in the
```
```suggestion
For the complex-valued case, the same formula as for the [`Sphere`](@ref) $𝕊^1$ is applied to values in the
```

sethaxen commented on src/manifolds/Euclidean.jl:
```diff
@@ -54,125 +54,125 @@ Compute the Euclidean distance between two points on the [`Euclidean`](@ref)
 Compute the inner product on the [`Euclidean`](@ref) `M`, which is just
 the inner product on the real-valued or complex valued vector space
-of arrays (or tensors) of size $n_1\times n_2 \times \cdots \times n_i$, i.e.
+of arrays (or tensors) of size $n_1 \times n_2  \times  …  \times n_i$, i.e.
```
Would `×` have the wrong spacing here?

kellertuer replied:
No, replaced them, too. I somehow missed too often to replace `\times`. Have done that with a global replace now.

sethaxen commented on src/manifolds/Euclidean.jl:
`````diff
@@ -54,125 +54,125 @@ Compute the Euclidean distance between two points on the [`Euclidean`](@ref)
+g_p(X,Y) = \sum_{k ∈ I} \overline{X}_{k} Y_{k},
 ````
-where $I$ is the set of integer vectors $k\in\mathbb N^i$, such that for all
+where $I$ is the set of integer vectors $k ∈ ℕ^i$, such that for all
`````
Integers are usually denoted with `ℤ`.

kellertuer replied:
then we should rephrase this since the inequalities directly after even restrict the natural numbers even further, maybe we can just leave out the word `integer` since the vectors k in ... is clear enough?

sethaxen commented on src/manifolds/Euclidean.jl:
```diff
@@ -230,63 +230,63 @@ function median(
+    project_point(M::Euclidean, p)
 
-Project an arbitrary point `x` onto the [`Euclidean`](@ref) `M`, which
+Project an arbitrary point `p` onto the [`Euclidean`](@ref) `M`, which
```
```suggestion
Project an arbitrary point `p` onto the [`Euclidean`](@ref) manifold `M`, which
```

sethaxen commented on src/manifolds/FixedRankMatrices.jl:
```diff
@@ -1,39 +1,40 @@
     FixedRankMatrices{m,n,k,T} <: Manifold
 
-The manifold of $m\times n$ real-valued (complex-valued) matrices of fixed rank $k$, i.e.
+The manifold of $m \times n$ real-valued or complex-valued matrices of fixed rank $k$, i.e.
```
Can `×` go here and below?

kellertuer replied:
Ah yes; I also checked for all other occurrences of `\times` and replaced them with `×`.

sethaxen commented on src/manifolds/GraphManifold.jl:
```diff
@@ -46,84 +46,83 @@ end
-the number of edges (for [`EdgeManifold`](@ref)s) and that each element of `v`
-together with its corresponding einty of `x` passes the
+the number of edges (for [`EdgeManifold`](@ref)s) and that each element of `X`
+together with its corresponding einty of `p` passes the
```
einty?

kellertuer replied:
should've been `entry` I don't know how one can mistype it like that.

sethaxen commented on src/manifolds/Grassmann.jl:
```diff
@@ -51,223 +49,235 @@ The manifold is named after
-            eltype(x),
-            "The matrix $(x) is neiter real- nor complex-valued matrix, so it does noe lie on the complex Grassmann manifold of dimension ($(n),$(k)).",
+            eltype(p),
+            "The matrix $(p) is neiter real- nor complex-valued matrix, so it does noe lie on the complex Grassmann manifold of dimension ($(n),$(k)).",
```
```suggestion
            "The matrix $(p) is neither a real- nor complex-valued matrix, so it does not lie on the complex Grassmann manifold of dimension ($(n),$(k)).",
```

sethaxen commented on src/manifolds/Grassmann.jl:
```diff
@@ -51,223 +49,235 @@ The manifold is named after
-            eltype(x),
-            "The matrix $(x) is not a real-valued matrix, so it does noe lie on the Grassmann manifold of dimension ($(n),$(k)).",
+            eltype(p),
+            "The matrix $(p) is not a real-valued matrix, so it does noe lie on the Grassmann manifold of dimension ($(n),$(k)).",
```
```suggestion
            "The matrix $(p) is not a real-valued matrix, so it does not lie on the Grassmann manifold of dimension ($(n),$(k)).",
```

sethaxen commented on src/manifolds/Grassmann.jl:
```diff
@@ -51,223 +49,235 @@ The manifold is named after
-            size(x),
-            "The matrix $(x) is does not lie on the Grassmann manifold of dimension ($(n),$(k)), since its dimensions are wrong.",
+            size(p),
+            "The matrix $(p) is does not lie on the Grassmann manifold of dimension ($(n),$(k)), since its dimensions are wrong.",
```
```suggestion
            "The matrix $(p) does not lie on the Grassmann manifold of dimension ($(n),$(k)), since its dimensions are wrong.",
```

sethaxen commented on src/manifolds/Grassmann.jl:
```diff
@@ -51,223 +49,235 @@ The manifold is named after
-            eltype(v),
-            "The matrix $(v) is neiter real- nor complex-valued matrix, so it can not bea tangent vector to the complex Grassmann manifold of dimension ($(n),$(k)).",
+            eltype(X),
+            "The matrix $(X) is neiter real- nor complex-valued matrix, so it can not bea tangent vector to the complex Grassmann manifold of dimension ($(n),$(k)).",
```
```suggestion
            "The matrix $(X) is neither a real- nor complex-valued matrix, so it can not be a tangent vector to the complex Grassmann manifold of dimension ($(n),$(k)).",
```

sethaxen commented on src/manifolds/Grassmann.jl:
```diff
@@ -51,223 +49,235 @@ The manifold is named after
-            size(v),
-            "The matrix $(v) is does not lie in the tangent space of $(x) on the Grassmann manifold of dimension ($(n),$(k)), since its dimensions are wrong.",
+            size(X),
+            "The matrix $(X) is does not lie in the tangent space of $(p) on the Grassmann manifold of dimension ($(n),$(k)), since its dimensions are wrong.",
```
```suggestion
            "The matrix $(X) does not lie in the tangent space of $(p) on the Grassmann manifold of dimension ($(n),$(k)), since its dimensions are wrong.",
```

sethaxen commented on src/manifolds/Grassmann.jl:
```diff
@@ -51,223 +49,235 @@ The manifold is named after
-            norm(x' * v + v' * x),
-            "The matrix $(v) is does not lie in the tangent space of $(x) on the Grassmann manifold of dimension ($(n),$(k)), since x'v + v'x is not the zero matrix.",
+            norm(p' * X + X' * p),
+            "The matrix $(X) is does not lie in the tangent space of $(p) on the Grassmann manifold of dimension ($(n),$(k)), since x'v + v'x is not the zero matrix.",
```
```suggestion
            "The matrix $(X) does not lie in the tangent space of $(p) on the Grassmann manifold of dimension ($(n),$(k)), since x'v + v'x is not the zero matrix.",
```

sethaxen commented on src/manifolds/MetricManifold.jl:
```diff
@@ -37,44 +37,44 @@ giving the signature $(++...+-)$.
+$\Gamma_{ijk} = \frac{1}{2} \Bigl[g_{kj,i} + g_{ik,j} - g_{ij,k}\Bigr],$
 
-where $g_{ij,k}=\frac{\partial}{\partial x^k} g_{ij}$ is the coordinate
+where $g_{ij,k}=\frac{\partial}{\partial p^k} g_{ij}$ is the coordinate
```
Can we use `∂`?

kellertuer replied:
Yes, good idea. I replaced all occurrences accordingly.

sethaxen commented on src/manifolds/MetricManifold.jl:
`````diff
@@ -114,127 +114,127 @@ function christoffel_symbols_second_jacobian(M::MetricManifold, x; backend = :de
 
 ````math
-w^\flat = G_xw,
+X^\flat = G_pX,
`````
```suggestion
X^♭= G_p X,
```

sethaxen commented on src/manifolds/MetricManifold.jl:
`````diff
@@ -276,182 +276,182 @@ function _convert_with_default(M::MT, T::Type{<:Metric}, ::Val{false}) where {MT
 
 ````math
-g_x(v,w) = \langle v, G_x w\rangle,
+g_p(X, Y) = ⟨X, G_pY⟩,
`````
```suggestion
g_p(X, Y) = ⟨X, G_p Y⟩,
```

sethaxen commented on src/manifolds/MetricManifold.jl:
```diff
@@ -276,182 +276,182 @@ function _convert_with_default(M::MT, T::Type{<:Metric}, ::Val{false}) where {MT
+    log_local_metric_density(M::MetricManifold, p)
 
-Return the natural logarithm of the metric density $\rho$ of `M` at `x`, which
+Return the natural logarithm of the metric density $\rho$ of `M` at `p`, which
```
```suggestion
Return the natural logarithm of the metric density $ρ$ of `M` at `p`, which
```
here and below

sethaxen commented on src/manifolds/MetricManifold.jl:
`````diff
@@ -464,76 +464,76 @@ function projected_distribution(M::MMT, ::Val{false}, d) where {MMT<:MetricManif
 
 ````math
-w^\sharp = G_x^{-1}w,
+ξ^\sharp = G_p^{-1}ξ,
`````
```suggestion
ξ^♯ = G_p^{-1} ξ,
```

sethaxen commented on src/manifolds/MetricManifold.jl:
```diff
@@ -544,7 +544,7 @@ Approximate the exponential map on the manifold over the provided timespan
 equation
 
-$\frac{d^2}{dt^2} x^k + \Gamma^k_{ij} \frac{d}{dt} x_i \frac{d}{dt} x_j = 0,$
+$\frac{d^2}{dt^2} p^k + \Gamma^k_{ij} \frac{d}{dt} p_i \frac{d}{dt} p_j = 0,$
```
```suggestion
$\frac{d^2}{dt^2} p^k + Γ^k_{ij} \frac{d}{dt} p_i \frac{d}{dt} p_j = 0,$
```

sethaxen commented on src/manifolds/MetricManifold.jl:
```diff
@@ -86,26 +86,26 @@ where $\Gamma_{ijk}$ are the Christoffel symbols of the first kind, and
-function christoffel_symbols_second(M::MetricManifold, x; backend = :default)
-    ginv = inverse_local_metric(M, x)
-    Γ₁ = christoffel_symbols_first(M, x; backend = backend)
+function christoffel_symbols_second(M::MetricManifold, p; backend = :default)
```
Can you use `Γ` instead of `\Gamma` above?

kellertuer replied:
Of course. I replaced all occurrences accordingly.

sethaxen commented on src/manifolds/MetricManifold.jl:
```diff
@@ -544,7 +544,7 @@ Approximate the exponential map on the manifold over the provided timespan
-$\frac{d^2}{dt^2} x^k + \Gamma^k_{ij} \frac{d}{dt} x_i \frac{d}{dt} x_j = 0,$
+$\frac{d^2}{dt^2} p^k + \Gamma^k_{ij} \frac{d}{dt} p_i \frac{d}{dt} p_j = 0,$
 
 where $\Gamma^k_{ij}$ are the Christoffel symbols of the second kind, and
```
```suggestion
where $Γ^k_{ij}$ are the Christoffel symbols of the second kind, and
```

sethaxen commented on src/manifolds/PowerManifold.jl:
```diff
@@ -230,107 +230,107 @@ end
-This can be done elementwise, so r every entry of `w` (and `x`) sparately
+use the musical isomorphism to transform the tangent vector `X` from the tangent space at
+`p` on an [`AbstractPowerManifold`](@ref) `M` to a cotangent vector.
+This can be done elementwise, so r every entry of `X` (and `p`) sparately
```
```suggestion
This can be done elementwise for each entry of `X` (and `p`).
```

sethaxen commented on src/manifolds/PowerManifold.jl:
```diff
@@ -404,154 +402,154 @@ function injectivity_radius(M::AbstractPowerManifold, x)
-$\mathcal M$, the manifold is of dimension
+$=𝒩 = (ℳ)^{n_1,…,n_d}, i.e. with $n=(n_1,…,n_d)$ the array
+size of the power manifold and $d_{ℳ}$ the dimension of the base manifold
+$ℳ$, the manifold is of dimension
```
use `\dim` below

kellertuer replied:
Ah yes, we should use `\dim` more consistently.

sethaxen commented on src/manifolds/ProductManifold.jl:
```diff
@@ -103,41 +103,41 @@ function PrecomputedProductOrthonormalBasis(
-Check whether `v` is a tangent vector to `x` on the [`ProductManifold`](@ref)
-`M`, i.e. atfer [`check_manifold_point`](@ref)`(M, x)`, and all projections to
+Check whether `X` is a tangent vector to `p` on the [`ProductManifold`](@ref)
+`M`, i.e. atfer [`check_manifold_point`](@ref)`(M, p)`, and all projections to
```
```suggestion
`M`, i.e. after [`check_manifold_point`](@ref)`(M, p)`, and all projections to
```

sethaxen commented on src/manifolds/ProductManifold.jl:
```diff
@@ -181,139 +181,139 @@ function cross(M1::ProductManifold, M2::ProductManifold)
+    distance(M::ProductManifold, p, q)
 
-compute the distance two points `x` and `y` on the [`ProductManifold`](@ref) `M`, which is
+compute the distance two points `p` and `q` on the [`ProductManifold`](@ref) `M`, which is
```
```suggestion
Compute the distance between two points `p` and `q` on the [`ProductManifold`](@ref) `M`, which is
```

sethaxen commented on src/manifolds/ProductManifold.jl:
```diff
@@ -181,139 +181,139 @@ function cross(M1::ProductManifold, M2::ProductManifold)
+    exp(M::ProductManifold, p, X)
 
-compute the exponential map from `x` towards `v` on the [`ProductManifold`](@ref) `M`,
+compute the exponential map from `p` towards `X` on the [`ProductManifold`](@ref) `M`,
```
```suggestion
compute the exponential map from `p` in the direction of `X` on the [`ProductManifold`](@ref) `M`,
```

sethaxen commented on src/manifolds/ProductManifold.jl:
```diff
@@ -181,139 +181,139 @@ function cross(M1::ProductManifold, M2::ProductManifold)
 
 use the musical isomorphism to transform the tangent vector `w` from the tangent space at
 `x` on the [`ProductManifold`](@ref) `M` to a cotangent vector.
 This can be done elementwise, so for every entry of `w` (and `x`) sparately
```
```suggestion
This can be done elementwise for every entry of `w` (and `x`) separately.
```

kellertuer commented:
Uh when I type things, especially short notes fast in English, the German in my breaks through and messes up capitalisation – I will check that and add your suggestions to the table. And will go through your remarks.

kellertuer commented on docs/src/manifolds/torus.md:
```diff
@@ -1,11 +1,11 @@
+The torus $𝕋^d \equiv [-π,π)^d$ is modeled as an [`AbstractPowerManifold`](@ref) of
 the (real-valued) [`Circle`](@ref) and uses [`MultidimentionalArrayPowerRepresentation`](@ref).
-Points on the torus are hence row vectors, $x\in\mathbb R^{d}$.
+Points on the torus are hence row vectors, $x\inℝ^{d}$.
```
While within a table it seems to break math regularly, here – yes of course. I will also check for remaining `\in`s (I hope I already have most).

kellertuer commented on docs/src/manifolds/torus.md:
```diff
@@ -1,11 +1,11 @@
 # Torus
-The torus $\mathbb T^d \equiv [-π,π)^d$ is modeled as an [`AbstractPowerManifold`](@ref) of
+The torus $𝕋^d \equiv [-π,π)^d$ is modeled as an [`AbstractPowerManifold`](@ref) of
```
Yes it does, but it's still the wrong symbol. I often confuse `\equiv` and `\cong`, which is the right one here.

kellertuer commented on src/Manifolds.jl:
`````diff
@@ -108,15 +108,15 @@ Abstract type for defining statistical estimation methods.
 
 ````math
-\wedge: v^i \mapsto v^i e_i
+\wedge: v^i ↦ v^i e_i
`````
...and similarly (if you do not note that below) `⋁` for `\vee`.

kellertuer commented:
Thanks for so carefully revising my proposed unification. It was actually some work; after finishing here I will to the same to base, also introducing code blue there'n stuff. I think I already reworked all your changes – noticed that the proposed changes (thanks for directly doing those) can be batch-commited when going through them in the files changed tab :)

Just two questions remain I think
> * Our differential notation from `Rotations` and the group manifolds
> * Our translation notation `τ` from group manifolds

Can you elaborate on those two? I added `\circ` to the notations. Is that what you meant with the first? The `τ` I could not find within notations for group manifolds.

sethaxen commented:
> Thanks for so carefully revising my proposed unification.

Thanks for doing this!

> Just two questions remain I think
> 
> > * Our differential notation from `Rotations` and the group manifolds
> > * Our translation notation `τ` from group manifolds
> 
> Can you elaborate on those two? I added `\circ` to the notations. Is that what you meant with the first? The `τ` I could not find within notations for group manifolds.

`\circ` in the notations is nice. There are also other operations that could go there in addition to the inner product, e.g. `\times` as a Cartesian product.

What I meant here was the differential in the sense used for `translate_diff` and `apply_diff`. It would be nice if we gave a more precise definition, but I think here it's just important to note the notation.

Also `τ` appears as a generalization of `L` and `R`, the left or right translations. It appears in
- https://julianlsolvers.github.io/Manifolds.jl/latest/manifolds/group.html#Manifolds.inverse_translate_diff-Tuple{Manifold,Any,Any,Any}
- https://julianlsolvers.github.io/Manifolds.jl/latest/manifolds/group.html#Manifolds.translate_diff-Tuple{Manifold,Any,Any,Any}
- https://julianlsolvers.github.io/Manifolds.jl/latest/manifolds/group.html#Manifolds.apply_diff-Tuple{AbstractGroupAction,Any,Any,Any}
- https://julianlsolvers.github.io/Manifolds.jl/latest/manifolds/group.html#Manifolds.inverse_apply_diff-Tuple{AbstractGroupAction,Any,Any,Any}

but it is used extensively in docstrings in #96. We should probably use it more extensively in `translate`, `inverse_translate`, `apply`, etc. If you'd prefer, I could handle this docstring update in #96 once this is merged.

I'll continue the review where I left off yesterday.

sethaxen commented on src/manifolds/ProductManifold.jl:
```diff
@@ -521,62 +522,62 @@ end
-This can be done elementwise, so vor every entry of `w` (and `x`) sparately
+Use the musical isomorphism to transform the cotangent vector `ξ` from the tangent space at
+`p` on the [`ProductManifold`](@ref) `M` to a tangent vector.
+This can be done elementwise, so vor every entry of `ξ` (and `p`) sparately
```
```suggestion
This can be done elementwise for every entry of `ξ` (and `p`) separately
```

sethaxen commented on src/manifolds/Rotations.jl:
```diff
@@ -1,14 +1,15 @@
-real-valued orthogonal matrices with determinant $+1$.
+The special orthogonal manifold $\mathrm{SO}(n)$ represented by $n × n$
+real-valued orthogonal matrices with determinant $+1$ is the manifold of `Rotations`,
+since these matrices represent all rotations in $ℝ^n$.
```
```suggestion
since these matrices represent all rotations of points in $ℝ^n$.
```

sethaxen commented on src/manifolds/Rotations.jl:
```diff
@@ -37,12 +38,12 @@ function NormalRotationDistribution(
-The Lie algebra of $\mathrm{SO}(4)$ consists of 4x4 skew-symmetric matrices.
-The unique imaginary components of their eigenvalues are the angles of the two
-plane rotations. This function computes these more efficiently than
+The Lie algebra of [`Rotations`](@ref) in $ℝ^4$, $\mathrm{SO}(4)$, consists of $4× 4$
```
```suggestion
The Lie algebra of [`Rotations(4)`](@ref) in $ℝ^{4 × 4}$, $𝔰𝔬(4)$, consists of $4 × 4$
```

sethaxen commented on src/manifolds/Rotations.jl:
`````diff
@@ -137,67 +137,69 @@ function cos_angles_4d_rotation_matrix(R)
-`v`, i.e.
 ````math
-\exp_xv = x \operatorname{Exp}(v),
+\exp_pX = p \operatorname{Exp}(X),
`````
```suggestion
\exp_p X = p \operatorname{Exp}(X),
```

sethaxen commented on src/manifolds/Rotations.jl:
```diff
@@ -230,25 +232,33 @@ function exp!(M::Rotations{4}, y, x, v)
+@doc raw"""
+    hat(M::Rotations, p, ω)
+
+Convert the unique tangent vector components $\omega$ at point `p` on [`Rotations`](@ref)
```
While the notation `ω` and `Ω` for `SO(n)` is common due to its use for angular velocity in physics, we should probably here adopt the same notation we use elsewhere, that is `Xⁱ` and `X`.

sethaxen commented on src/Manifolds.jl:
`````diff
@@ -108,15 +108,15 @@ Abstract type for defining statistical estimation methods.
 
 ````math
-\wedge: v^i \mapsto v^i e_i
+∧ : v^i ↦ v^i e_i
`````
Let's adopt the tangent vector notation `X` and coefficients `X^i` here and for the variable names. Also for `vee`.

kellertuer replied:
Yes, I basically just missed to rename these two functions parameters.

sethaxen commented on src/manifolds/Rotations.jl:
```diff
@@ -230,25 +232,33 @@ function exp!(M::Rotations{4}, y, x, v)
-    v² = v * v
-    y .= a₀ .* x .+ x * (a₁ .* v .+ a₂ .* v² .+ a₃ .* (v² * v))
-    return y
+    v² = X * X
```
```suggestion
    X² = X * X
```

sethaxen commented on src/manifolds/Rotations.jl:
```diff
@@ -230,25 +232,33 @@ function exp!(M::Rotations{4}, y, x, v)
-    y .= a₀ .* x .+ x * (a₁ .* v .+ a₂ .* v² .+ a₃ .* (v² * v))
-    return y
+    v² = X * X
+    q .= a₀ .* p .+ p * (a₁ .* X .+ a₂ .* v² .+ a₃ .* (v² * X))
```
```suggestion
    q .= a₀ .* p .+ p * (a₁ .* X .+ a₂ .* X² .+ a₃ .* (X² * X))
```

sethaxen commented on src/manifolds/ProductManifold.jl:
```diff
@@ -655,31 +656,31 @@ support(d::ProductPointDistribution) = MPointSupport(d.manifold)
         SM = first(t)
         dim = manifold_dimension(SM)
-        tvⁱ = @inbounds view(vⁱ, i:(i+dim-1))
+        tvⁱ = @inbounds view(Xⁱ, i:(i+dim-1))
```
```suggestion
        tXⁱ = @inbounds view(Xⁱ, i:(i+dim-1))
```

sethaxen commented on src/manifolds/ProductManifold.jl:
```diff
@@ -655,31 +656,31 @@ support(d::ProductPointDistribution) = MPointSupport(d.manifold)
         dim = manifold_dimension(SM)
-        tvⁱ = @inbounds view(vⁱ, i:(i+dim-1))
+        tvⁱ = @inbounds view(Xⁱ, i:(i+dim-1))
         vee!(SM, tvⁱ, Base.tail(t)...)
```
```suggestion
        vee!(SM, tXⁱ, Base.tail(t)...)
```

sethaxen commented on src/manifolds/ProductManifold.jl:
```diff
@@ -335,172 +335,173 @@ function get_vectors(
     for t ∈ ts
         dim = manifold_dimension(first(t))
-        tvⁱ = @inbounds view(vⁱ, i:(i+dim-1))
+        tvⁱ = @inbounds view(Xⁱ, i:(i+dim-1))
```
```suggestion
        tXⁱ = @inbounds view(Xⁱ, i:(i+dim-1))
```

sethaxen commented on src/manifolds/ProductManifold.jl:
```diff
@@ -335,172 +335,173 @@ function get_vectors(
         dim = manifold_dimension(first(t))
-        tvⁱ = @inbounds view(vⁱ, i:(i+dim-1))
+        tvⁱ = @inbounds view(Xⁱ, i:(i+dim-1))
         hat!(t..., tvⁱ)
```
```suggestion
        hat!(t..., tXⁱ)
```

sethaxen commented on src/manifolds/ProductManifold.jl:
```diff
@@ -181,139 +181,139 @@ function cross(M1::ProductManifold, M2::ProductManifold)
-@doc doc"""
-    flat(M::ProductManifold, x, w::FVector{TangentSpaceType})
+@doc raw"""
+    flat(M::ProductManifold, p, w::FVector{TangentSpaceType})
```
This docstring needs an update

sethaxen commented on src/manifolds/Rotations.jl:
```diff
@@ -466,27 +471,27 @@ Compute the Riemannian [`mean`](@ref mean(M::Manifold, args...)) of `x` using
 
-i.e. the Frobenius norm of `v`, where tangent vectors are represented by
+i.e. the Frobenius norm of `X`, where tangent vectors are represented by
 elements from the Lie group.
```
```suggestion
elements from the Lie algebra.
```

sethaxen commented on src/manifolds/Rotations.jl:
```diff
@@ -505,78 +510,79 @@ $D_{ij}=\begin{cases} \operatorname{sgn}(R_{ij}) & \text{if} \; i=j \\ 0 & \, \t
+Project `p` to the nearest point on manifold `M`.
 
-Given the singular value decomposition $x = U \Sigma V^\mathrm{T}$, with the
+Given the singular value decomposition $p = U \Sigma V^\mathrm{T}$, with the
```
```suggestion
Given the singular value decomposition $p = U Σ V^\mathrm{T}$, with the
```

sethaxen commented on src/manifolds/Rotations.jl:
```diff
@@ -608,50 +614,52 @@ function _fix_random_rotation(A::AbstractMatrix)
+    retract(M, p, X, ::QRRetraction)
+
+Compute the SVD-based retraction on the [`Rotations`](@ref) `M` from `p` in direction `X`
+(as an element of the Lie group) and is a first-order approximation of the exponential map.
```
```suggestion
(as an element of the Lie group), which is a first-order approximation of the exponential map.
```

sethaxen commented on src/manifolds/Rotations.jl:
```diff
@@ -608,50 +614,52 @@ function _fix_random_rotation(A::AbstractMatrix)
+@doc raw"""
+    retract(M, p, X, ::QRRetraction)
+
+Compute the SVD-based retraction on the [`Rotations`](@ref) `M` from `p` in direction `X`
```
```suggestion
Compute the QR-based retraction on the [`Rotations`](@ref) `M` from `p` in direction `X`
```

sethaxen commented on src/manifolds/Sphere.jl:
```diff
@@ -1,14 +1,14 @@
     Sphere(n)
 
-Generate the $\mathbb S^{n}\subset \mathbb R^{n+1}$
+Generate the $𝕊^{n}\subset ℝ^{n+1}$
```
```suggestion
Generate $𝕊^{n} ⊂ ℝ^{n+1}$.
```

sethaxen commented on src/manifolds/Sphere.jl:
```diff
@@ -17,187 +17,187 @@ Sphere(n::Int) = Sphere{n}()
 function get_basis(M::Sphere{N}, x, B::DiagonalizingOrthonormalBasis) where {N}
```
Variable needs an update

sethaxen commented on src/manifolds/Sphere.jl:
`````diff
@@ -17,187 +17,187 @@ Sphere(n::Int) = Sphere{n}()
 
 ````math
-d_{\mathbb S^n}(x,y) = \operatorname{acos}(\langle x, y\rangle).
+d_{𝕊^n}(p,q) = \operatorname{acos}(⟨p,q⟩).
`````
Should we use `\arccos`, since it has a LaTeX command?

sethaxen commented on src/manifolds/Sphere.jl:
```diff
@@ -17,187 +17,187 @@ Sphere(n::Int) = Sphere{n}()
 
-Compute the exponential map from `x` into the tangent direction `v` on the [`Sphere`](@ref)
-`M` by following the great arc eminating from `x` in direction `v` along with length of `v`.
+Compute the exponential map from `p` into the tangent direction `X` on the [`Sphere`](@ref)
```
```suggestion
Compute the exponential map from `p` in the tangent direction `X` on the [`Sphere`](@ref)
```

sethaxen commented on src/manifolds/Sphere.jl:
```diff
@@ -17,187 +17,187 @@ Sphere(n::Int) = Sphere{n}()
+function get_vector(M::Sphere{N}, p, X, B::ArbitraryOrthonormalBasis) where {N}
+    p[1] ≈ 1 && return vcat(0, X)
+    xp1 = p .+ ntuple(i -> ifelse(i == 1, 1, 0), N + 1)
+    v0 = vcat(0, X)
```
Here and below change `v0` to `X0`

sethaxen commented on src/manifolds/Stiefel.jl:
```diff
@@ -30,263 +30,281 @@ The manifold is named after
-is (approximatly) the identity, where $\cdot^{\mathrm{H}}$ is the complex conjugate
-transpose. The settings for approximately can be set with `kwargs...`.
+Check whether `p` is a valid point on the [`Stiefel`](@ref) `M`=$\operatorname{St}(n,k)$, i.e. that it has the right
+[`AbstractNumbers`](@ref) type and $p^{\mathrm{H}}p$ is (approximatly) the identity, where $\cdot^{\mathrm{H}}$ is the
```
```suggestion
[`AbstractNumbers`](@ref) type and $p^{\mathrm{H}}p$ is (approximately) the identity, where $\cdot^{\mathrm{H}}$ is the
```

sethaxen commented on src/manifolds/Stiefel.jl:
```diff
@@ -30,263 +30,281 @@ The manifold is named after
 `M`=$\operatorname{St}(n,k)$, i.e. the [`AbstractNumbers`](@ref) fits and
-it (approximtly) holds that $x^{\mathrm{H}}v + v^{\mathrm{H}}x = 0$, where
-`kwargs...` is passed to the `isapprox`.
+it (approximtly) holds that $p^{\mathrm{H}}X + X^{\mathrm{H}}p = 0$.
```
```suggestion
it (approximately) holds that $p^{\mathrm{H}}X + X^{\mathrm{H}}p = 0$.
```

sethaxen commented on src/manifolds/Stiefel.jl:
```diff
@@ -30,263 +30,281 @@ The manifold is named after
-Compute the exponential map on the [`Stiefel`](@ref)`{n,k,T}`() manifold `M`
-eminating from `x` into tangent direction `v`.
+Computes the exponential map on the [`Stiefel`](@ref)`{n,k,T}`() manifold `M`
+eminating from `p` into tangent direction `X`.
```
```suggestion
emanating from `p` in tangent direction `X`.
```

sethaxen commented on src/manifolds/Symmetric.jl:
```diff
@@ -1,243 +1,245 @@
 [`AbstractNumbers`](@ref) `F`.
 
-Though it is slighty redundant, usually the matrices are safed as $n\times n$ arrays.
+Though it is slighty redundant, usually the matrices are safed as $n × n$ arrays.
```
```suggestion
Though it is slightly redundant, usually the matrices are stored as $n × n$ arrays.
```

sethaxen commented on src/manifolds/Symmetric.jl:
```diff
@@ -1,243 +1,245 @@
-Check whether `v` is a tangent vector to manifold point `x` on the
-[`SymmetricMatrices`](@ref) `M`, i.e. `v` has to be a symmetric matrix of dimension `(n,n)`
+Check whether `X` is a tangent vector to manifold point `p` on the
+[`SymmetricMatrices`](@ref) `M`, i.e. `X` has to be a symmetric matrix of dimension `(n,n)`
```
```suggestion
[`SymmetricMatrices`](@ref) `M`, i.e. `X` has to be a symmetric matrix of size `(n,n)`
```

sethaxen commented on src/manifolds/Symmetric.jl:
```diff
@@ -1,243 +1,245 @@
     dim = manifold_dimension(M)
-    vout = similar(v, dim)
-    @assert size(v) == (N, N)
+    vout = similar(X, dim)
```
`vout` -> `Xout` or `Y` here and below

sethaxen commented on src/manifolds/Symmetric.jl:
```diff
@@ -253,96 +255,96 @@ function manifold_dimension(::SymmetricMatrices{N,𝔽}) where {N,𝔽}
 
-@doc doc"""
+@doc raw"""
     sharp(M::SymmetricMatrices, x, w::FVector{CotangentSpaceType})
```
Needs a notation update

sethaxen commented on src/manifolds/SymmetricPositiveDefiniteLinearAffine.jl:
```diff
@@ -8,74 +8,74 @@ struct LinearAffineMetric <: RiemannianMetric end
 ) where {N}
-    xSqrt = sqrt(x)
-    eigv = eigen(B.v)
+    xSqrt = sqrt(p)
```
x -> p

sethaxen commented on src/manifolds/SymmetricPositiveDefiniteLogCholesky.jl:
`````diff
@@ -10,143 +10,143 @@ introduced by [^Lin2019].
 ````
-where $l$ is the colesky factor of $x$ and $w=\log_lk$ for $k$ the cholesky factor
-of $y$ and the just mentioned logarithmic map is the one on [`CholeskySpace`](@ref).
+where $x$ is the colesky factor of $p$ and $W=\log_xy$ for $y$ the cholesky factor
`````
```suggestion
where $x$ is the cholesky factor of $p$ and $W=\log_x y$ for $y$ the cholesky factor
```

sethaxen commented on src/orthonormal_bases.jl:
```diff
@@ -70,21 +70,21 @@ function ProjectedOrthonormalBasis(method::Symbol, F::AbstractNumbers = ℝ)
 An orthonormal basis `Ξ` as a vector of tangent vectors (of length determined by
 [`manifold_dimension`](@ref)) in the tangent space that diagonalizes the curvature
-tensor $R(u,v)w$ and where the direction `v` has curvature `0`.
+tensor $R(u,v)w$ and where the direction `frameDirection` $v$ has curvature `0`.
```
```suggestion
tensor $R(u,v)w$ and where the direction `frame_direction` $v$ has curvature `0`.
```

sethaxen commented on src/statistics.jl:
```diff
@@ -33,7 +33,7 @@ t_k &= \frac{w_k}{\sum_{i=1}^k w_i}\\
 where $x_k$ are points, $w_k$ are weights, $\mu_k$ is the $k$th estimate of the
```
`\mu` -> `μ` throughout this file

sethaxen commented on src/statistics.jl:
```diff
@@ -365,32 +365,32 @@ function mean!(
+    yold = allocate_result(M, median, q)
     ytmp = copy(yold)
-    v = zero_tangent_vector(M, y)
+    v = zero_tangent_vector(M, q)
```
`v` to `X` in this file. Probably useful to change `x` to `p` as elsewhere to avoid confusion.

kellertuer replied:
I left `x` when it was a vector of points on the manifold. but I changed `x0` to `p0`.

kellertuer commented on docs/src/notation.md:
```diff
@@ -0,0 +1,33 @@
+| $\mathcal M$ | a manifold | $\mathcal M_1, \mathcal M_2,\ldots,\mathcal N$ | |
+| $\operatorname{Exp}$ | the matrix exponential | |
+| $\operatorname{Log}$ | the matrix logarithm | |
+| $\mathcal P_{q\gets p}X$ | parallel transport | | of the vector $X$ from $T_p\mathcal M$ to $T_q\mathcal M$
```
Thats easy, if you concatenate parallel transports $P_{q\gets p}P_{p\gets s}X$, i.e. you transport X from s -> p -> q this version is far better readable than $P_{p\to q}P_{s\to p}$ since for this the order is obscured. The first one can be read just from right to left.

kellertuer commented:
Additionally to $\times$ I also added $\hat$ for the Cartesian (n-ary) power.

It would be great if you could add $\tau$ within your #96 then, when you use
it from then on more often anyways.

I tried to find further `v`-resemblemses not that I also did not (xet) rename
all tests and their variables.

For the matrices I also prefer non-bold non-mathrm, just $0_k$ and $I_k$ and added that to the notations as well.

Basis vectors are just $\Xi = \{X_1,\ldots,X_n\}$, i.e. $\Xi$ is a set of vectors, maybe a basis. I thought we use $X_1$ and $X^1$ more like depending on the Einstein summation? This might not yet be fully unified.

Finally, I don't know how I came to "eminating" I corrected those occurences.

I will keep this one open a day or two just to give everybody a possibility to take a final look.

mateuszbaran commented on docs/src/notation.md:
```diff
@@ -0,0 +1,41 @@
+| $^{\wedge}$ | (n-ary) Cartesian power of a manifold | | see [`PowerManifold`](@ref) |
+| $T^*_p \mathcal M$ | the cotangent space at $p$ | | |
+| $\xi$ | a cotangent vector from $T^*_p \mathcal M$ | $\xi_1, \xi_2,\ldots,\eta,\zeta$ | sometimes written with base point $\xi_p$. |
+| $n$ | dimension (of a manifold) | $n_1,n_2,\ldots,m, \operatorname{dim}(\mathcal M)$| for the real dimension sometimes also $\operatorname{dim}_{\mathbb R}(\mathcal M)$|
```
In docstrings `\dim` is used instead of `\operatorname{dim}`. Is this due to a documentation rendering issue?

kellertuer replied:
It's mainly due to me having missed to replace those, after I noticed `\dim` is available. I replaced those two, too, now.

mateuszbaran commented on src/manifolds/PowerManifold.jl:
```diff
@@ -607,21 +605,21 @@ method has to be one that is available on the base [`Manifold`](@ref).
+@doc raw"""
     sharp(M::AbstractPowerManifold, x, w::FVector{CotangentSpaceType})
 
 Use the musical isomorphism to transform the cotangent vector `w` from the tangent space at
```
This docstring uses the old notation.

kellertuer replied:
Thanks for noticing.
